### PR TITLE
bakery: define new encoding format for third party caveats

### PR DESCRIPTION
**Pull request by rogpeppe**
Instead of a string, we will use a protobuf encoding
that will include the need-declared information as
a field and the actual condition body as a binary
byte slice.

Note that the .pb.go files at automatically generated.

### CODE REVIEW

**Review by alesstimec — APPROVED** [review]

File: bakery/authorizer_test.go [hunk 1]
```diff
@@ -27,14 +27,12 @@ func (*authorizerSuite) TestAuthorizerFunc(c *gc.C) {
 		case "b":
 			return true, nil, nil
 		case "c":
-			return true, []checkers.Caveat{{
-				Location:  "somewhere",
-				Condition: "c",
+			return true, []checkers.Caveat{{Location: "somewhere",
```
Comment (mhilton) [inline bakery/authorizer_test.go:30]: This seems an odd formatting, why change it.

File: bakery/authorizer_test.go [hunk 2]
```diff
@@ -27,14 +27,12 @@ func (*authorizerSuite) TestAuthorizerFunc(c *gc.C) {
...
+				ThirdPartyCondition: []byte("c"),
 			}}, nil
 		case "d":
-			return true, []checkers.Caveat{{
-				Location:  "somewhere",
-				Condition: "d",
+			return true, []checkers.Caveat{{Location: "somewhere",
```
Comment (mhilton) [inline bakery/authorizer_test.go:34]: and this one

File: bakery/checkers/checkers.go [hunk 3]
```diff
@@ -169,18 +169,39 @@ func checkError(ctx context.Context, _, arg string) error {
...
 
 // Caveat represents a condition that must be true for a check to
-// complete successfully. If Location is non-empty, the caveat must be
-// discharged by a third party at the given location.
-// The Namespace field holds the namespace URI of the
-// condition - if it is non-empty, it will be converted to
-// a namespace prefix before adding to the macaroon.
+// complete successfully. It can represent either a first-party
+// or a third-party caveat.
 type Caveat struct {
```
Comment (mhilton) [inline bakery/checkers/checkers.go:174]: It seems to me that there is something wrong with having a type where all the fields a mutually exclusive (no field is shared between first and third party caveats). Perhaps this should be two types with some kind of common interface?

**Review by mhilton — APPROVED** [review]
LGTM with a few thoughts

File: httpbakery/client_test.go [hunk 4]
```diff
@@ -1271,10 +1271,10 @@ func newDischargeRequiredError(hp serverHandlerParams, checkErr error, req *http
 	}
 	var caveats []checkers.Caveat
 	if hp.authLocation != "" {
-		caveats = []checkers.Caveat{{
-			Location:  hp.authLocation,
-			Condition: "is-ok",
+		caveats = []checkers.Caveat{{Location: hp.authLocation,
```
Comment (mhilton) [inline httpbakery/client_test.go:1274]: Fromatting again

File: httpbakery/client_test.go [hunk 5]
```diff
@@ -423,10 +423,10 @@ func (s *ClientSuite) serverRequiringMultipleDischarges(n int, discharger *baker
 			w.Write([]byte("ok"))
 			return
 		}
-		caveats := []checkers.Caveat{{
-			Location:  discharger.Location(),
-			Condition: "is-ok",
+		caveats := []checkers.Caveat{{Location: discharger.Location(),
```
Comment (mhilton) [inline httpbakery/client_test.go:426]: And here

File: httpbakery/agent/agent_test.go [hunk 6]
```diff
@@ -461,9 +461,8 @@ func (c idmClient) IdentityFromContext(ctx context.Context) (bakery.Identity, []
 }
 
 func identityCaveats(dischargerURL string) []checkers.Caveat {
-	return []checkers.Caveat{{
-		Location:  dischargerURL,
-		Condition: "test condition",
+	return []checkers.Caveat{{Location: dischargerURL,
```
Comment (mhilton) [inline httpbakery/agent/agent_test.go:464]: and here

**Comment by rogpeppe** [comment]
This is ancient and it's really superceded by the fact that third party caveat information can now be stored externally, so size isn't such a big concern any more.

**Pull request closed without merging**